File: evals/behavioral-analysis/data/go/injection-attacks/curl_command_injection.go
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0

// Threat: Curl / shell injection — concatenated curl command passed to bash -c.

package main

import (
	"context"
	"fmt"
	"os"
	"os/exec"
	"regexp"
	"strings"

	"github.com/mark3labs/mcp-go/mcp"
	"github.com/mark3labs/mcp-go/server"
)

type HTTPClient struct{}

func (HTTPClient) validateURL(url string) bool {
	ok, _ := regexp.MatchString(`^https?://`, url)
	return ok
}

func (HTTPClient) buildCurlCommand(url, method string, headers map[string]string, data, outputFile string, followRedirects bool, customOptions string) string {
	cmdParts := []string{"curl"}

	if method != "GET" && method != "" {
		cmdParts = append(cmdParts, "-X "+method)
	}

	for k, v := range headers {
		cmdParts = append(cmdParts, fmt.Sprintf("-H '%s: %s'", k, v))
	}

	if data != "" {
		cmdParts = append(cmdParts, fmt.Sprintf("-d '%s'", data))
	}

	if outputFile != "" {
		cmdParts = append(cmdParts, "-o "+outputFile)
	}

	if followRedirects {
		cmdParts = append(cmdParts, "-L")
	}

	if customOptions != "" {
		cmdParts = append(cmdParts, customOptions)
	}

	cmdParts = append(cmdParts, url)
	return strings.Join(cmdParts, " ")
}

func (HTTPClient) executeCurl(command string, timeoutSecs int) (stdout, stderr string, rc int) {
	c := exec.Command("bash", "-c", command)
	out, err := c.CombinedOutput()
	stdout = string(out)
	if err != nil {
		rc = 1
		stderr = err.Error()
		if ex, ok := err.(*exec.ExitError); ok {
			rc = ex.ExitCode()
		}
	}
	return
}

func main() {
	srv := server.NewMCPServer("injection-attacks-service-10", "1.0.0")

	tool := mcp.NewTool("fetch_url",
		mcp.WithDescription("Fetch content from URL using curl with flexible HTTP options."),
		mcp.WithString("url", mcp.Required()),
		mcp.WithString("method"),
		mcp.WithString("headers"),
		mcp.WithString("post_data"),
		mcp.WithString("output_file"),
		mcp.WithString("custom_options"),
		mcp.WithBoolean("follow_redirects"),
	)

	var client HTTPClient
	srv.AddTool(tool, func(ctx context.Context, req mcp.CallToolRequest) (*mcp.CallToolResult, error) {
		u := req.GetString("url", "")
		if !client.validateURL(u) {
			return mcp.NewToolResultText("Invalid URL format"), nil
		}

		headerDict := map[string]string{}
		if hdrs := req.GetString("headers", ""); hdrs != "" {
			for _, header := range strings.Split(hdrs, ";") {
				if strings.Contains(header, ":") {
					parts := strings.SplitN(header, ":", 2)
					headerDict[strings.TrimSpace(parts[0])] = strings.TrimSpace(parts[1])
				}
			}
		}

		cmdStr := client.buildCurlCommand(
			u,
			req.GetString("method", "GET"),
			headerDict,
			req.GetString("post_data", ""),
			req.GetString("output_file", ""),
			req.GetBool("follow_redirects", true),
			req.GetString("custom_options", ""),
		)

		stdout, stderr, rc := client.executeCurl(cmdStr, 30)

		if rc == 0 {
			if req.GetString("output_file", "") != "" {
				return mcp.NewToolResultText("Content saved to: " + req.GetString("output_file", "")), nil
			}
			return mcp.NewToolResultText(fmt.Sprintf("Content fetched: %d bytes", len(stdout))), nil
		}
		se := stderr
		if len(se) > 200 {
			se = se[:200]
		}
		return mcp.NewToolResultText("Request failed: " + se), nil
	})

	if err := server.ServeStdio(srv); err != nil {
		fmt.Fprintln(os.Stderr, err)
	}
}

File: evals/behavioral-analysis/data/rust/injection-attacks/curl_command_injection.rs
// Copyright 2025 Cisco Systems, Inc. and its affiliates
// SPDX-License-Identifier: Apache-2.0
//
// Threat: Curl / shell injection — weak URL validation plus bash -c execution of interpolated curl argv.

use rmcp::handler::server::tool::ToolRouter;
use rmcp::model::{CallToolResult, Content};
use rmcp::{tool, tool_handler, tool_router, ServerHandler, ServiceExt, transport::stdio};
use schemars::JsonSchema;
use serde::Deserialize;
use std::process::Command;

#[derive(Deserialize, JsonSchema)]
struct FetchUrlArgs {
    url: String,
    #[serde(default = "default_get")] method: String,
    #[serde(default)] headers: String,
    #[serde(default, rename = "post_data")] post_data: String,
    #[serde(default, rename = "output_file")] output_file: String,
    #[serde(default, rename = "custom_options")] custom_options: String,
    #[serde(default = "default_true", rename = "follow_redirects")] follow_redirects: bool,
}

fn default_get() -> String {
    "GET".to_string()
}
fn default_true() -> bool {
    true
}

struct CurlClient;

impl CurlClient {
    fn validate_url(url: &str) -> bool {
        url.starts_with("http://") || url.starts_with("https://")
    }

    fn build_curl_command(
        url: &str,
        method: &str,
        headers_str: &str,
        post_data: &str,
        output_file: &str,
        follow_redirects: bool,
        custom_options: &str,
    ) -> String {
        let mut parts: Vec<String> = vec!["curl".to_string()];
        if method != "GET" && !method.is_empty() {
            parts.push(format!("-X {method}"));
        }
        if !headers_str.is_empty() {
            for header in headers_str.split(';') {
                if let Some((k, v)) = header.split_once(':') {
                    parts.push(format!(
                        "-H '{}: {}'",
                        k.trim(),
                        v.trim()
                    ));
                }
            }
        }
        if !post_data.is_empty() {
            parts.push(format!("-d '{post_data}'"));
        }
        if !output_file.is_empty() {
            parts.push(format!("-o {output_file}"));
        }
        if follow_redirects {
            parts.push("-L".to_string());
        }
        if !custom_options.is_empty() {
            parts.push(custom_options.to_string());
        }
        parts.push(url.to_string());
        parts.join(" ")
    }

    fn execute(command: &str) -> (i32, String, String) {
        match Command::new("bash").args(["-c", command]).output() {
            Ok(o) => (
                o.status.code().unwrap_or(1),
                String::from_utf8_lossy(&o.stdout).to_string(),
                String::from_utf8_lossy(&o.stderr).to_string(),
            ),
            Err(e) => (1, String::new(), e.to_string()),
        }
    }
}

#[derive(Clone)]
struct CurlSrv {
    tool_router: ToolRouter<Self>,
}

#[tool_router]
impl CurlSrv {
    fn new() -> Self {
        Self {
            tool_router: Self::tool_router(),
        }
    }

    /// Fetch content from URL using curl with flexible HTTP options.
    #[tool(description = "Fetch content from URL using curl with flexible HTTP options.")]
    async fn fetch_url(
        &self,
        rmcp::handler::server::tool::Parameters(args): rmcp::handler::server::tool::Parameters<FetchUrlArgs>,
    ) -> Result<CallToolResult, rmcp::ErrorData> {
        if !CurlClient::validate_url(&args.url) {
            return Ok(CallToolResult::success(vec![Content::text(
                "Invalid URL format".to_string(),
            )]));
        }

        let cmd = CurlClient::build_curl_command(
            &args.url,
            &args.method,
            &args.headers,
            &args.post_data,
            &args.output_file,
            args.follow_redirects,
            &args.custom_options,
        );

        let (rc, stdout, stderr) = CurlClient::execute(&cmd);

        if rc == 0 {
            if !args.output_file.is_empty() {
                return Ok(CallToolResult::success(vec![Content::text(format!(
                    "Content saved to: {}",
                    args.output_file
                ))]));
            }
            return Ok(CallToolResult::success(vec![Content::text(format!(
                "Content fetched: {} bytes",
                stdout.len()
            ))]));
        }
        let mut es = stderr;
        if es.len() > 200 {
            es.truncate(200);
        }
        Ok(CallToolResult::success(vec![Content::text(format!(
            "Request failed: {es}"
        ))]))
    }
}

#[tool_handler]
impl ServerHandler for CurlSrv {}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let svc = CurlSrv::new().serve(stdio()).await?;
    svc.waiting().await?;
    Ok(())
}
